### PR TITLE
numpy: Handle multidimentional indexing in 1.15

### PR DESCRIPTION
**Pull request by hmaarrfk**
* Handles all numpy deprecations.
  * ~~Goes in an catches the warnings from dependencies that don't follow numpy 1.15~~ PR #3242.
  * ~~Locks the expected warnings to a specific version.~~ #3242
  * ~~A few other small fixes where `expected_warnings` was passed as string as opposed to a list.~~ #3242
  * Issue #3235 is handled as a single commit so that it can be rebased if we don't like it.

Future me notes: use
```bash
git checkout master
git pull upstream master
git rebase -i master
```
to pull it out and edit it to whatever the final solution is chosen to be.

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
[If this is a bug-fix or enhancement, it closes issue # ]
[If this is a new feature, it implements the following paper: ]

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by hmaarrfk** [comment]
Maybe a better way to handle scipy's use of the [matrix subclass](https://github.com/scipy/scipy/issues/8868#issuecomment-391974007) for sparse matricies is to use https://github.com/pydata/sparse internally.

File: skimage/restoration/_cycle_spin.py [hunk 1]
```diff
@@ -155,5 +155,5 @@ def _run_one_shift(shift):
         # multithreaded via dask
         futures = [dask.delayed(_run_one_shift)(s) for s in all_shifts]
         mean = sum(futures) / len(futures)
-        mean = mean.compute(get=dask.threaded.get, num_workers=num_workers)
+        mean = mean.compute(scheduler='threading', num_workers=num_workers)
```
Comment (soupault) [inline skimage/restoration/_cycle_spin.py]: To be fixed via https://github.com/scikit-image/scikit-image/pull/3205

File: skimage/segmentation/boundaries.py [hunk 2]
```diff
@@ -26,7 +26,7 @@ def _find_boundaries_subpixel(label_img):
 
     label_img_expanded = np.zeros([(2 * s - 1) for s in label_img.shape],
                                   label_img.dtype)
-    pixels = [slice(None, None, 2)] * ndim
+    pixels = tuple([slice(None, None, 2)] * ndim)
```
Comment (soupault) [inline skimage/segmentation/boundaries.py]: This is just `pixels = (slice(None, None, 2), ) * ndim`

File: skimage/util/_regular_grid.py [hunk 3]
```diff
@@ -54,7 +54,7 @@ def regular_grid(ar_shape, n_points):
     sorted_dims = np.sort(ar_shape)
     space_size = float(np.prod(ar_shape))
     if space_size <= n_points:
-        return [slice(None)] * ndim
+        return tuple([slice(None)] * ndim)
```
Comment (soupault) [inline skimage/util/_regular_grid.py]: Same :)

File: skimage/util/_regular_grid.py [hunk 4]
```diff
@@ -68,7 +68,7 @@ def regular_grid(ar_shape, n_points):
     stepsizes = np.round(stepsizes).astype(int)
     slices = [slice(start, None, step) for
               start, step in zip(starts, stepsizes)]
-    slices = [slices[i] for i in unsort_dim_idxs]
+    slices = tuple([slices[i] for i in unsort_dim_idxs])
```
Comment (soupault) [inline skimage/util/_regular_grid.py]: Why not `tuple(slices[i] for i in unsort_dim_idxs)`?

File: skimage/util/arraycrop.py [hunk 5]
```diff
@@ -169,7 +169,8 @@ def crop(ar, crop_width, copy=False, order='K'):
     """
     ar = np.array(ar, copy=False)
     crops = _validate_lengths(ar, crop_width)
-    slices = [slice(a, ar.shape[i] - b) for i, (a, b) in enumerate(crops)]
+    slices = tuple([slice(a, ar.shape[i] - b)
```
Comment (soupault) [inline skimage/util/arraycrop.py]: Same

File: skimage/util/shape.py [hunk 6]
```diff
@@ -248,7 +248,7 @@ def view_as_windows(arr_in, window_shape, step=1):
 
     arr_in = np.ascontiguousarray(arr_in)
 
-    slices = [slice(None, None, st) for st in step]
+    slices = tuple([slice(None, None, st) for st in step])
```
Comment (soupault) [inline skimage/util/shape.py]: Same

File: skimage/_shared/_warnings.py [hunk 7]
```diff
@@ -103,10 +103,16 @@ def expected_warnings(matching):
     Finally, you can use `|\A\Z` in a pattern to signify it as optional.
 
     """
+    if isinstance(matching, str):
```
Comment (soupault) [inline skimage/_shared/_warnings.py]: This is nice!

File: skimage/filters/lpi_filter.py [hunk 8]
```diff
@@ -20,7 +20,7 @@ def _centre(x, oshape):
 
     """
     start = (np.array(x.shape) - np.array(oshape)) // 2 + 1
-    out = x[[slice(s, s + n) for s, n in zip(start, oshape)]]
+    out = x[tuple([slice(s, s + n) for s, n in zip(start, oshape)])]
```
Comment (soupault) [inline skimage/filters/lpi_filter.py]: Also here, and below

File: skimage/filters/tests/test_lpi_filter.py [hunk 9]
```diff
@@ -1,12 +1,20 @@
...
 from skimage._shared import testing
+from distutils.version import LooseVersion as Version
 import unittest
 
 from skimage import data
 from skimage.filters import LPIFilter2D, inverse, wiener
+from skimage._shared._warnings import expected_warnings
 
 
+if Version(np.__version__) >= '1.15' and Version(scipy.__version__) <= '1.1.0':
```
Comment (soupault) [inline skimage/filters/tests/test_lpi_filter.py]: Are there any plans on removing this check in future? If yes, please, add a note to `TODO` file

File: skimage/restoration/tests/test_denoise.py [hunk 10]
```diff
@@ -342,23 +349,26 @@ def test_wavelet_denoising():
...
+            denoised = restoration.denoise_wavelet(noisy,
+                                                   multichannel=multichannel,
+                                                   convert2ycbcr=convert2ycbcr)
         psnr_noisy = compare_psnr(img, noisy)
         psnr_denoised = compare_psnr(img, denoised)
         assert_(psnr_denoised > psnr_noisy)
 
         # SNR is improved less with 1 wavelet level than with the default.
-        denoised_1 = restoration.denoise_wavelet(noisy,
+        with expected_warnings([None]):
```
Comment (soupault) [inline skimage/restoration/tests/test_denoise.py]: What is the benefit of using `None`?

File: skimage/util/_regular_grid.py (hunk 4 above)
Comment (hmaarrfk) [reply skimage/util/_regular_grid.py]: You know, I thought you needed the square brackets to make it a list comprehension. I guess you don't :D

Comment (soupault) [reply skimage/util/_regular_grid.py]: @hmaarrfk I hape a PEP for you :) - https://www.python.org/dev/peps/pep-0289/

File: skimage/restoration/tests/test_denoise.py (hunk 10 above)
Comment (hmaarrfk) [reply skimage/restoration/tests/test_denoise.py]: The benefit of using none is that under certain build conditions, where we don't expect warnings, `None` can be passed in.

**Comment by hmaarrfk** [comment]
@soupault plans for deprecacting the logic behind the conditional warnings:

Well, the plan for removing it is so long term that I didn't know if it makes sense. I'll add some words about it.

I assume that scipy (and pywavelets) will eventually fix these warnings on their side too. at which point, some version of scipy, higher than 1.1.6, will eventually not have these warnings by default. But until then, this is just a very explicit way ensuring warning are

File: skimage/restoration/_cycle_spin.py (hunk 1 above)
Comment (hmaarrfk) [reply skimage/restoration/_cycle_spin.py]: Ok. That was easy to remove using rebasing :D

**Comment by hmaarrfk** [comment]
Finally, I was testing out avoiding the use of `|\A\Z` since it isn't really explicit as to what conditions might trigger the warning.

Even without instructions, somebody 10 years from now would be able to go through the code, find that a certain combinations of packages can't exist in their supported versions, and rip out the logic if they want to.

That said, I don't know when pywavelets and scipy will address their warnings, maybe it is for us to send in the PR, and the tests might fail once or twice until they get around to fixing their codebase. We would require version bumps on the logic at that point.

The matrix subclass warning should definitely be broadened, as that would require a major change on scipy's end.

**Comment by hmaarrfk** [comment]
I just updated the scipy dependency to not depend on the minor version number for the sparse matrix warning.

File: skimage/util/_regular_grid.py (hunk 4 above)
Comment (hmaarrfk) [reply skimage/util/_regular_grid.py]: nice one! Yeah, I still don't know generator expressions.

**Comment by jni** [comment]
Before anyone merges, I suggested in https://github.com/scikit-image/scikit-image/issues/3235#issuecomment-402862546 that changing the return type of `rectangular_grid` should be done over a deprecation cycle.

**Comment by hmaarrfk** [comment]
@jni I finally got around to removing the keyword argument and added your suggested `Notes` section.

**Comment by hmaarrfk** [comment]
@jni I tried:
```
conda install sphinx-gallery scikit-learn sphinx numpydocs
sudo apt-get install texlive texlive-latex-extra dvipng  # big download
```
I got pretty far in the building process, but still no cigar. Something about a db mismatch.

**Comment by soupault** [comment]
@jni Looks good to me:
image

**Comment by soupault** [comment]
Thanks a lot @hmaarrfk !

**Comment by soupault** [comment]
@meeseeksdev backport to v0.14.x

**Comment by soupault** [comment]
Apparently, I have 0 knowledge on the manual backporting of merge commits... Could someone try to handle this, please?

**Comment by Carreau** [comment]
> Apparently, I have 0 knowledge on the manual backporting of merge commits... Could someone try to handle this, please?

That's the kind of issue you can raise on MrMeeseeks repo. The "there seem to be a conflict" comments should now have instruction on which git commands to issue to backport a merge commit

**Comment by hmaarrfk** [comment]
I just learned where MrMeeseeks is from....

**Comment by jni** [comment]
@hmaarrfk 

> I just learned where MrMeeseeks is from....

Time for some Netflix binge-watching. =D

**Pull request merged**